### PR TITLE
Use float constants for calculations with floats in TColor

### LINKED ISSUE #6373: Floating point exceptions in TColor class.
## Describe the bug

Floating point exception on OSX 10.15 with XCode11 (and others).
```
317.486847 427.758362 53.922855
212.467392 0.529984 94.476196

 *** Break *** floating point exception
[/usr/lib/system/libsystem_platform.dylib] _sigtramp (no debug info)
[<unknown binary>] (no debug info)
[/Users/user/Projects/xxx/root/lib/libCore.6.22.so] TColor::SetRGB(float, float, float) (no debug info)
[/Users/user/./color] main /Users/user/color.cxx:89
[/usr/lib/system/libdyld.dylib] start (no debug info)
[<unknown binary>] (no debug info)
```

## Expected behavior
No FPEs

## To Reproduce

Code:
```
//
// main.cxx

#include <iostream>
#include <stdlib.h>

#include <TColor.h>
#include <TRandom.h>
#include <fenv.h>

#ifndef HAVE_FEENABLEEXCEPT
#if defined(__APPLE__) && defined(__MACH__)

// Public domain polyfill for feenableexcept on OS X
// http://www-personal.umich.edu/~williams/archive/computation/fe-handling-example.c

inline int feenableexcept(unsigned int excepts)
{
    static fenv_t fenv;
    unsigned int new_excepts = excepts & FE_ALL_EXCEPT;
    // previous masks
    unsigned int old_excepts;

    if (fegetenv(&fenv)) {
        return -1;
    }
    old_excepts = fenv.__control & FE_ALL_EXCEPT;

    // unmask
    fenv.__control &= ~new_excepts;
    fenv.__mxcsr   &= ~(new_excepts << 7);

    return fesetenv(&fenv) ? -1 : old_excepts;
}

inline int fedisableexcept(unsigned int excepts)
{<!--
    static fenv_t fenv;
    unsigned int new_excepts = excepts & FE_ALL_EXCEPT;
    // all previous masks
    unsigned int old_excepts;

    if (fegetenv(&fenv)) {
        return -1;
    }
    old_excepts = fenv.__control & FE_ALL_EXCEPT;

    // mask
    fenv.__control |= new_excepts;
    fenv.__mxcsr   |= new_excepts << 7;

    return fesetenv(&fenv) ? -1 : old_excepts;
}

#else
inline int feenableexcept(unsigned int excepts)
{
    #pragma STDC FENV_ACCESS ON
    fexcept_t flags;
    /* Save current exception flags. */
    fegetexceptflag(&flags, FE_ALL_EXCEPT);

    feclearexcept(FE_ALL_EXCEPT);   /* clear all fp exception conditions */
    return fesetexceptflag(&flags, excepts) != 0 ? -1 : flags; /* set new flags */

}

inline int fedisableexcept(unsigned int excepts)
{
    #pragma STDC FENV_ACCESS ON
    fexcept_t flags;
    /* Save current exception flags. */
    fegetexceptflag(&flags, FE_ALL_EXCEPT);

    feclearexcept(FE_ALL_EXCEPT);   /* clear all fp exception conditions */
    return fesetexceptflag(&flags, ~excepts) != 0 ? -1 : flags; /* set new flags */
}

#endif
#endif


int main(int argc, char **argv) {
    feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
    TColor* c= new TColor();
    TRandom* r= new TRandom();
    float rgb[3];
    //c->SetRGB(212.467392, 0.529984, 94.476196);
    for (size_t i=0; i<1000000;i++)
    {
    rgb[0]=512*(1.0-r->Rndm());
    rgb[1]=512*(1.0-r->Rndm());
    rgb[2]=512*(1.0-r->Rndm());
    printf("%f %f %f\n",rgb[0],rgb[1],rgb[2]);
    c->SetRGB(rgb[0],rgb[1],rgb[2]);
    }
    delete c;
    delete r;
    return 0;
}
```
Compile
```
clang++ color.cxx $(root-config --glibs --cflags --libs) -o color
```
Run
```
./color
```
## Setup
1. ROOT 6.22.02 from the official site that matches XCode.

## Additional context
This is similar to #6344  but seen only on Mac.

### PR DESCRIPTION
**Pull request by andriish**
In this PR the calculations in TColor are done explicitly using float constants.

This is a continuation of #6677

The PR should fix #6373  

Tag @couet

### CODE REVIEW

**Comment by bellenot** [comment]
@phsft-bot build

**Comment by couet** [comment]
@phsft-bot build

**Comment by pcanal** [comment]
@phsft-bot build

**Comment by andriish** [comment]
Hi @couet ,
the reproducer/test  is

```
//
// main.cxx

#include <iostream>
#include <stdlib.h>

#include <TColor.h>
#include <TRandom.h>
#include <fenv.h>

#ifndef HAVE_FEENABLEEXCEPT
#if defined(__APPLE__) && defined(__MACH__)

// Public domain polyfill for feenableexcept on OS X
// http://www-personal.umich.edu/~williams/archive/computation/fe-handling-example.c

inline int feenableexcept(unsigned int excepts)
{
    static fenv_t fenv;
    unsigned int new_excepts = excepts & FE_ALL_EXCEPT;
    // previous masks
    unsigned int old_excepts;

    if (fegetenv(&fenv)) {
        return -1;
    }
    old_excepts = fenv.__control & FE_ALL_EXCEPT;

    // unmask
    fenv.__control &= ~new_excepts;
    fenv.__mxcsr   &= ~(new_excepts << 7);

    return fesetenv(&fenv) ? -1 : old_excepts;
}

inline int fedisableexcept(unsigned int excepts)
{
    static fenv_t fenv;
    unsigned int new_excepts = excepts & FE_ALL_EXCEPT;
    // all previous masks
    unsigned int old_excepts;

    if (fegetenv(&fenv)) {
        return -1;
    }
    old_excepts = fenv.__control & FE_ALL_EXCEPT;

    // mask
    fenv.__control |= new_excepts;
    fenv.__mxcsr   |= new_excepts << 7;

    return fesetenv(&fenv) ? -1 : old_excepts;
}

#else
inline int feenableexcept(unsigned int excepts)
{
    #pragma STDC FENV_ACCESS ON
    fexcept_t flags;
    /* Save current exception flags. */
    fegetexceptflag(&flags, FE_ALL_EXCEPT);

    feclearexcept(FE_ALL_EXCEPT);   /* clear all fp exception conditions */
    return fesetexceptflag(&flags, excepts) != 0 ? -1 : flags; /* set new flags */

}

inline int fedisableexcept(unsigned int excepts)
{
    #pragma STDC FENV_ACCESS ON
    fexcept_t flags;
    /* Save current exception flags. */
    fegetexceptflag(&flags, FE_ALL_EXCEPT);

    feclearexcept(FE_ALL_EXCEPT);   /* clear all fp exception conditions */
    return fesetexceptflag(&flags, ~excepts) != 0 ? -1 : flags; /* set new flags */
}

#endif
#endif


int main(int argc, char **argv) {
    feenableexcept(FE_DIVBYZERO | FE_INVALID | FE_OVERFLOW);
    TRandom* R1= new TRandom();
    TRandom* R2= new TRandom();
    TRandom* R3= new TRandom();
    Float_t r1,g1,b1;
    Float_t r2,g2,b2;
    Float_t r3,g3,b3;
    Float_t fHue1=-1.0f; 
    Float_t fLight1=-1.0f;
    Float_t fSaturation1 = -1.0f;
    Float_t fHue2=-1.0f; 
    Float_t fLight2=-1.0f;
    Float_t fSaturation2 = -1.0f;
    Float_t fHue3=-1.0f; 
    Float_t fLight3=-1.0f;
    Float_t fSaturation3 = -1.0f;
    for (size_t i=0; i<1000000;i++)
    {
    r1=16.0f*(1.0f-float(R1->Rndm()));
    g1=16.0f*(1.0f-float(R1->Rndm()));
    b1=16.0f*(1.0f-float(R1->Rndm()));
    printf("TColor::RGB2HLS(%f, %f, %f ...)->",r1,g1,b1);
    TColor::RGB2HLS( r1,g1,b1 ,fHue1, fLight1, fSaturation1);
    printf("%f, %f, %f\n",fHue1, fLight1, fSaturation1);
    }
    for (size_t i=0; i<1000000;i++)
    {
    r2=16.0f*(1.0f-float(R2->Rndm()));
    g2=16.0f*(1.0f-float(R2->Rndm()));
    b2=16.0f*(1.0f-float(R2->Rndm()));
    printf("TColor::RGB2HSV(%f, %f, %f ...) ->",r2,g2,b2);
    TColor::RGB2HSV( r2,g2,b2 ,fHue2, fLight2, fSaturation2);
    printf("%f, %f, %f\n",fHue2, fLight2, fSaturation2);
    }
    for (size_t i=0; i<1000000;i++)
    {
    r3=16.0f*(1.0f-float(R3->Rndm()));
    g3=16.0f*(1.0f-float(R3->Rndm()));
    b3=16.0f*(1.0f-float(R3->Rndm()));
    printf("TColor::HLStoRGB(%f, %f, %f ...) ->",r3,g3,b3);
    TColor::HLStoRGB( r3,g3,b3 ,fHue3, fLight3, fSaturation3);
    printf("%f, %f, %f\n",fHue3, fLight3, fSaturation3);
    }
    
    
    delete R1;
    delete R2;
    delete R3;
    return 0;
}
```
One should compile it with ``clang++  -O4``.
On Mac  + official ROOT 6.22 it gives me a crash almost immediately 
```
TColor::RGB2HLS(13.531198, 9.228077, 3.350407 ...)->0.000000, 1.000000, 0.000000
TColor::RGB2HLS(12.057813, 12.900666, 13.290458 ...)->0.000000, 1.000000, 0.000000
TColor::RGB2HLS(3.584785, 12.519324, 2.208079 ...)->0.000000, 1.000000, 0.000000
TColor::RGB2HLS(6.409876, 8.155184, 1.691731 ...)->0.000000, 1.000000, 0.000000

 *** Break *** floating point exception
TColor::RGB2HLS(9.862289, 0.100378, 3.557429 ...)->[/usr/lib/system/libsystem_platform.dylib] _sigtramp (no debug info)

```
However it runs fine on Linux where ROOT was build with gcc.

It seems the other functions do not crash.  Or , at least, I was not able to find proper parameters.


Best regards,

Andrii

**Comment by couet** [comment]
@andriish I was thinking of a test on the modele of those in core/base/test

**Comment by andriish** [comment]
Hi @couet ,

hm. That is google test. I don't use it. Will try to adapt the reproducer to it, but it will take some time.

Best regards,

Andrii

**Comment by couet** [comment]
I guess a quick test on the various XXX2YYY methods is enough (input value xyz should gives output abc...) no need for a reproducer. Other wise do not worry we can add it later.

**Review by couet — APPROVED** [review]
It looks good.

File: core/base/src/TColor.cxx [hunk 1]
```diff
@@ -1605,33 +1605,33 @@ void TColor::RGB2HLS(Float_t rr, Float_t gg, Float_t bb,
    if (g > maxval) maxval = g;
    if (b > maxval) maxval = b;
 
-   rnorm = gnorm = bnorm = 0;
+   rnorm = gnorm = bnorm = 0.0f;
```
Comment (Axel-Naumann) [inline core/base/src/TColor.cxx:1608]: Not needed

File: core/base/src/TColor.cxx [hunk 2]
```diff
@@ -1605,33 +1605,33 @@ void TColor::RGB2HLS(Float_t rr, Float_t gg, Float_t bb,
...
    msum  = maxval + minval;
-   light = 0.5 * msum;
+   light = 0.5f * msum;
    if (maxval != minval) {
       rnorm = (maxval - r)/mdiff;
       gnorm = (maxval - g)/mdiff;
       bnorm = (maxval - b)/mdiff;
    } else {
-      satur = hue = 0;
+      satur = hue = 0.0f;
```
Comment (Axel-Naumann) [inline core/base/src/TColor.cxx:1617]: No effect.

File: core/base/src/TColor.cxx [hunk 3]
```diff
@@ -1660,21 +1660,21 @@ void TColor::RGB2HSV(Float_t r, Float_t g, Float_t b,
    if (max != 0) {
       satur = delta/max;
    } else {
-      satur = 0;
-      hue   = -1;
+      satur = 0.0f;
+      hue   = -1.0f;
```
Comment (Axel-Naumann) [inline core/base/src/TColor.cxx:1664]: No effect

**Review by Axel-Naumann — CHANGES_REQUESTED** [review]
I am not convinced that we need these changes. Do you have an example of what they fix? Much of this has no effect on the resulting assembly; they are simply code churn. Some might result in slightly different assembly instructions, but I double they have a visible effect - can you convince us that this is needed?

I would be good if you could expand the commit log to convey that info, too, by including the case that this change fixes.

**Comment by andriish** [comment]
Hi @Axel-Naumann ,

> Much of this has no effect on the resulting assembly; they are simply code churn. Some might result in slightly different assembly instructions, but I double they have a visible effect - can you convince us that this is needed?

This would be completely true if the clang would do its job well. Not the case, unfortunately. Therefore  

> slightly different assembly instructions

actually matter. 


> by including the case that this change fixes.

The reproducer should not crash with these changes.


>I don't see why this is needed.

What I actually did is that I've put   'f' everywhere for the sake of consistency.
I completely agree that at least in some cases the changes will not matter even for clang. 

>I'd be in favor of reducing the change to the few lines where it actually has an effect.

OK. Revert the comparisons and leave the assignments? Does that sound reasonable?



Best regards,

Andrii

**Comment by Axel-Naumann** [comment]
> The reproducer should not crash with these changes.

Could you verify that, please?

> OK. Revert the comparisons and leave the assignments? Does that sound reasonable?

Even assignments should be left with double literals (`float x = 1.2;`); as you can see from the CompilerExplorer link I pasted above, clang does generate the exact same assembly whether you assign a float or double literal. Instead, only computations with literals should become floats, as these will otherwise become double precision computations. So based on my current understanding my preference would be:

- leave double literals in comparisons and assignments;
- change to float literals where they appearing in computations (`+ - * /`), if you can validate that this fixes your reproducer.

Would that sound reasonable?

**Comment by andriish** [comment]
Hi @Axel-Naumann ,

> So based on my current understanding my preference would be: ...

Done.

> Could you verify that, please?

The standalone example does not clash with floats.  To check the example with ROOT I will need some time. At least to compile ROOT with clang. 


Best regards,

Andrii

**Comment by andriish** [comment]
Hi @Axel-Naumann ,

just checked.

master branch with -O4:

```
host:build user$ source /Users/user/MYROOT/bin/thisroot.sh 
host:build user$ clang++ -O4 color.cxx  -o color.exe $(root-config --glibs --cflags)
clang: warning: -O4 is equivalent to -O3 [-Wdeprecated]
host:build user$ ./color.exe 
TColor::RGB2HLS(13.531198, 9.228077, 3.350407 ...)->0.000000, 1.000000, 0.000000
TColor::RGB2HLS(12.057813, 12.900666, 13.290458 ...)->0.000000, 1.000000, 0.000000
TColor::RGB2HLS(3.584785, 12.519324, 2.208079 ...)->0.000000, 1.000000, 0.000000
TColor::RGB2HLS(6.409876, 8.155184, 1.691731 ...)->0.000000, 1.000000, 0.000000

 *** Break *** floating point exception
TColor::RGB2HLS(9.862289, 0.100378, 3.557429 ...)->[/usr/lib/system/libsystem_platform.dylib] _sigtramp (no debug info)
[<unknown binary>] (no debug info)
[/Users/user/Projects/ROOT/build/./color.exe] main (no debug info)
[/usr/lib/system/libdyld.dylib] start (no debug info)
[<unknown binary>] (no debug info)
host:build user$ 
```

TColorClang branch with -O4 runs fine.
Not posting output for obvious reasons.



Best regards,

Andrii

**Comment by andriish** [comment]
Hi @couet , @Axel-Naumann ,

would you have a look? This PR was discussed with a lot of details...  Not having these fixes at least in nightlies prevents me from looking at other things in the debugged code. ( The stuff I debug simply crashes before ).  And I would really like to debug as much as possible so the results would go into 6.24.

Best regards,

Andrii

**Comment by couet** [comment]
looks fine for me.

**Review by Axel-Naumann — APPROVED** [review]

**Comment by andriish** [comment]
> @andriish I was thinking of a test on the model of those in core/base/test

Hi @couet , I've wrote that 

> Will try to adapt the reproducer to it, but it will take some time.

But so far I don't have any progress there.

**Pull request merged**